### ISSUE TITLE
Def.sequential(...) throws unexpected exception

### ISSUE THREAD

**Issue by kamilkloch** [sbt/sbt#2948]
Set up a dummy task:
```sbt
  lazy val dummyTask = taskKey[Unit]("Dummy task.")
  dummyTask := {()}
```
This works:
```sbt
    compile in Compile := Def.sequential(
      compile in Compile
    ).value
```

Adding a dummy task causes an exception:
```sbt
    compile in Compile := Def.sequential(
      dummyTask,
      compile in Compile
    ).value
```

```sbt
[error] setting(ScopedKey(Scope(Select(ProjectRef(file:/home/kamilk/gitlab/x-scala/,commons)),Select(ConfigKey(compile)),Global,Global),compile)) at LinePosition(/home/kamilk/gitlab/x-scala/project/ReleaseSettings.scala,30) referenced from ScopedKey(Scope(Select(ProjectRef(file:/home/kamilk/gitlab/x-scala/,commons)),Select(ConfigKey(compile)),Global,Global),compile)
```

**Comment by eed3si9n** [comment]
This is probably expected since recursive calls are not allowed for flatMap / dynamic tasks.

**Comment by jvican** [comment]
To me, this is a serious bug in the way sbt handles the references to previous task definitions and it's just a leak of the current implementation. The expectation as a user is that this would work. It's annoying that the internal task cannot be referenced when `dummyTask` is added to `Def.sequential` and it works without it. Examples such as this end up being one of the reasons why people heavily complain about sbt: whenever you think you've grokked sbt's model, sbt surprises you with an inconsistency. That is frustrating for users... it'd be nice if sbt fixes this behaviour in a future version.

**Comment by eed3si9n** [comment]
ok. So the expectation is that reassignment doesn't create an infinite loop, but instead call the older definition? I could reopen this issue.

**Comment by jvican** [comment]
Yes, that's the expectation. The access to the previous task definition should be always available.
